synapse/cryotank.py

```diff
@@ -571,3 +589,644 @@ def init(self, name, conf=None, timeout=None):
+        '''
+        Undo a pauseIndex.
+        Args:
+            prop: (Optional[str]):  the index to start indexing, or if None, indicate to resume all indices
+        Returns:
+            None
+        '''
+        return self._meta.resumeIndex(prop)
+
+    @_inWorker
+    def getIndices(self):
+        '''
+        Args:
+            None
+        Returns
+            List[Dict[str: Any]]: all the indices with progress and statistics
+        '''
+        idxs = {iid: dict(metaentry.asdict()) for iid, metaentry in self._meta.indices.items()}
+        for iid in idxs:
+            idxs[iid].update(self._meta.progresses.get(iid, {}))
+        return list(idxs.values())
+
+    def _iterrows(self, prop, valu, exact=False):
+        '''
+        Query against an index.
+
+        Args:
+            prop (str):  The name of the indexed property
+            valu (Optional[Union[int, str]]):  The normalized value.  If not present, all records with prop present,
+            sorted by prop will be returned.  It will be considered prefix if exact is False.
+            exact (bool): Indicates that the result must match exactly.  Conversly, if False, indicates a prefix match.
+
+        Returns:
+            Iterable[Tuple[int, bytes, bytes, lmdb.Transaction]: a generator of a Tuple of the offset, the encoded
+            offset, the encoded index ID, and the LMDB read transaction.
+
+        Note:
+            ordering of Tuples disregard everything after the first 128 bytes of a property.
+        '''
+        iid = self._meta.iidFromProp(prop)
+        if iid is None:
+            raise ValueError("%s isn't being indexed" % prop)
+        iidenc = _iid_en(iid)
+
+        islarge = valu is not None and isinstance(valu, str) and len(valu) >= s_lmdb.LARGE_STRING_SIZE
+        if islarge and not exact:
+            valu = valu[:s_lmdb.LARGE_STRING_SIZE]  # type: ignore
+
+        if islarge and exact:
+            key = iidenc + s_lmdb.encodeValAsKey(valu)
+        elif valu is None:
+            key = iidenc
+        else:
+            key = iidenc + s_lmdb.encodeValAsKey(valu, isprefix=not exact)
+        with self._dbenv.begin(db=self._idxtbl, buffers=True) as txn, txn.cursor() as curs:
+            if exact:
+                rv = curs.set_key(key)
+            else:
+                rv = curs.set_range(key)
+            if not rv:
+                return
+            while True:
+                rv = []
+                curkey, offset_enc = curs.item()
+                if (not exact and not curkey[:len(key)] == key) or (exact and curkey != key):
+                    return
+                offset = _Int64be.unpack(offset_enc)[0]
+                yield (offset, offset_enc, iidenc, txn)
+                if not curs.next():
+                    return
+
+    def normValuByPropVal(self, prop, valu=None, exact=False):
+        '''
+        Query for normalized individual property values.
+
+        Args:
+            See _iterrows
```

Comment:
Top level APIs should not refer to "internal" APIs for argument reference ( since they're more free to change and should be "black box" to the object user )  In general, it would be preferable to document the API in both places to make reference use easy.

Reply:
ok